### PR TITLE
IA-2637 QA-1489 Fix flakiness in pd mount

### PR DESCRIPTION
**Pull request by Qi77Qi**
\<your comments for this PR go here\>

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
```
15:06:38 failed to register layer: Error processing tar file(exit status 1): write /google-cloud-sdk/lib/third_party/grpc/_cython/cygrpc.so: no space left on device
15:06:38 Build step 'Execute shell' marked build as failure
```

jenkins retest

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
leonardo-build failed...

jenkins retest

**Comment by Qi77Qi** [comment]
multi-test

File: http/src/main/resources/init-resources/cloud-init.yml [hunk 1]
```diff
@@ -3,16 +3,10 @@
 bootcmd:
 - fsck.ext4 -tvy /dev/sdb
 - mkdir -p /mnt/disks/work
-- chmod a+rwx /mnt/disks/work
 - mount -t ext4 -O discard,defaults /dev/sdb /mnt/disks/work
+- chmod a+rwx /mnt/disks/work
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/cloud-init.yml]: moving chmod to after `mount`

File: http/src/main/resources/init-resources/gce-init.sh [hunk 2]
```diff
@@ -246,13 +246,6 @@ fi
 # done GCS copy
 STEP_TIMINGS+=($(date +%s))
 
-# If any image is hosted in a GCR registry (detected by regex) then
-# authorize docker to interact with gcr.io.
-if grep -qF "gcr.io" <<< "${JUPYTER_DOCKER_IMAGE}${RSTUDIO_DOCKER_IMAGE}${PROXY_DOCKER_IMAGE}${WELDER_DOCKER_IMAGE}" ; then
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/gce-init.sh:251]: this doesn't really work. but I think it's not needed anyway

File: http/src/main/resources/init-resources/gce-init.sh [hunk 3]
```diff
@@ -321,6 +314,9 @@ ${DOCKER_COMPOSE} --env-file=/var/variables.env "${COMPOSE_FILES[@]}" config
 
 retry 5 ${DOCKER_COMPOSE} --env-file=/var/variables.env "${COMPOSE_FILES[@]}" pull
 
+# This needs to happen before we start up containers
+chmod a+rwx ${WORK_DIRECTORY}
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/gce-init.sh:320]: this has to be before `docker-compose up`

File: http/src/main/resources/init-resources/startup.sh [hunk 4]
```diff
@@ -74,6 +74,25 @@ SERVER_CRT=$(proxyServerCrt)
 SERVER_KEY=$(proxyServerKey)
 ROOT_CA=$(rootCaPem)
 
+mkdir -p /mnt/disks/work
+chmod a+rwx /mnt/disks/work
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/startup.sh]: I think what happens is when runtime is rebooted, the directory is re-created, but permission sometimes is still not set properly (not entirely sure why sometimes things work..). Hence, adding chmod here as well to reduce some flakiness

File: http/src/main/resources/init-resources/startup.sh [hunk 5]
```diff
@@ -86,7 +105,6 @@ function failScriptIfError() {
 
 function validateCert() {
   certFileDirectory=$1
-  dockerCompose=$3
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/startup.sh:89]: pass the cmd as an argument doesn't really work. Hence moving the function down and referencing `DOCKER_COMPOSE` directly

**Review by rtitle — APPROVED** [review]
Changes look fine to me.. :shipit: if tests pass reliably. Thanks for looking into this

**Comment by Qi77Qi** [comment]
multi-test

previous comment didn't trigger anything

**Comment by rtitle** [comment]
I think `multi-test` needs to be its own comment

File: http/src/main/resources/init-resources/startup.sh [hunk 6]
```diff
@@ -74,9 +74,6 @@ SERVER_CRT=$(proxyServerCrt)
 SERVER_KEY=$(proxyServerKey)
 ROOT_CA=$(rootCaPem)
 
-mkdir -p /mnt/disks/work
```
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/startup.sh]: getting this in one of the test
```
e2fsck 1.46.2 (28-Feb-2021)
/dev/sdb is in use.
e2fsck: Cannot continue, aborting.


mount: /mnt/disks/work: /dev/sdb already mounted or mount point busy.
```

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
fiab start failure... jenkins restart

**Comment by Qi77Qi** [comment]
jenkins retest (not restart)

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
fiab start failure....jenkins retest

**Comment by Qi77Qi** [comment]
jenkins retest

File: http/src/main/resources/init-resources/startup.sh [hunk 7]
```diff
@@ -98,39 +113,32 @@ function validateCert() {
...
 
+### This is only needed for gce. But doesn't hurt to have it either for non gce VMs
+#sleep 5s
+#ls -la /mnt/disks/work
+#mkdir -p /mnt/disks/work
+#chmod a+rwx /mnt/disks/work
+
+fsck.ext4 -tvy /dev/sdb
+mkdir -p /mnt/disks/work
+mount -t ext4 -O discard,defaults /dev/sdb /mnt/disks/work
```
Comment (rtitle) [inline http/src/main/resources/init-resources/startup.sh]: this `mount` should probably not happen for Dataproc, right?

Comment (Qi77Qi) [reply http/src/main/resources/init-resources/startup.sh]: ah yea...I just wanted to see if this fixes the mount flakiness first for gce

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
multi-test

**Review by rtitle — APPROVED** [review]

**Comment by Qi77Qi** [comment]
random fiab test failure...jenkins retest

**Comment by Qi77Qi** [comment]
multi-test

**Comment by Qi77Qi** [comment]
fiab start failure in one of the multi test....all other test runs passed....going to merge

**Pull request merged**